plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/data/ExecuteInsertBatchImpl.java

```diff
@@ -160,52 +160,43 @@ StringBuilder prepareQueryForStatement(
         }
         query.append(")\n\tVALUES "); //$NON-NLS-1$
         StringBuilder valuesPart = new StringBuilder(64);
-        valuesPart.append("(");
-        hasKey = false;
         boolean skipBindValues = CommonUtils.toBoolean(options.get(DBSDataManipulator.OPTION_SKIP_BIND_VALUES));
-        for (int i = 0; i < attributes.length; i++) {
-            DBSAttributeBase attribute = attributes[i];
-            if (DBUtils.isPseudoAttribute(attribute) || (!useMultiRowInsert && (!allNulls && DBUtils.isNullValue(attributeValues[i])))) {
-                continue;
-            }
-            if (hasKey) valuesPart.append(","); //$NON-NLS-1$
-            hasKey = true;
-
-            DBDValueHandler valueHandler = handlers[i];
-            
-            if (valueHandler instanceof DBDValueBinder) {
-                valuesPart.append(((DBDValueBinder) valueHandler).makeQueryBind(attribute, attributeValues[i]));
-            } else {
-                valuesPart.append("?"); //$NON-NLS-1$
-            }
-        }
-        valuesPart.append(")"); //$NON-NLS-1$
-        if (useMultiRowInsert) {
-            for (int i = 0; i < attributeValues.length / attributes.length; i++) {
-                if (i != 0) {
-                    query.append(",");
+        int noOfRecords = attributeValues.length / attributes.length;
+        int noOfRecordsM1 = noOfRecords - 1;
+        int k = 0;
+        for (int j = 0; j < noOfRecords; j++) {
+            valuesPart.append("(");
+            hasKey = false;
+            for (int i = 0; i < attributes.length; i++, k++) {
+                DBSAttributeBase attribute = attributes[i];
+                if (DBUtils.isPseudoAttribute(attribute) || (!useMultiRowInsert && (!allNulls && DBUtils.isNullValue(attributeValues[k])))) {
+                    continue;
+                }
+                if (hasKey)
+                    valuesPart.append(","); //$NON-NLS-1$
+                hasKey = true;
+
+                DBDValueHandler valueHandler = handlers[i];
+
+                if (valueHandler instanceof DBDValueBinder) {
+                    valuesPart.append(((DBDValueBinder) valueHandler).makeQueryBind(attribute, attributeValues[i]));
+                } else {
+                    if (skipBindValues) {
+                        valuesPart.append(SQLUtils.convertValueToSQL(session.getDataSource(), attributes[i], valueHandler, attributeValues[k], DBDDisplayFormat.NATIVE));
+                    } else {
+                        valuesPart.append("?"); //$NON-NLS-1$
+                    }
                 }
-                query.append(valuesPart);
             }
-        } else {
```

Comment:
You can loop over `attributeValues` instead of `attributes` to bind values properly and also support `useMultiRowInsert`. Both `handler` and `attribute` can be received using `array[i % array.length]`.